From 841861e376a4dadb810113ac9dc553185ea6d853 Mon Sep 17 00:00:00 2001
From: Kaifeng Zhu <kaifeng.zhu@citrix.com>
Date: Fri, 7 Mar 2014 15:03:40 +0000
Subject: [PATCH 02/10] block-cow: don't close cow_fd twice on error
To: andrew.cooper3@citrix.com

Coverity issue #25099

Signed-off-by: Kaifeng Zhu <kaifeng.zhu@citrix.com>
diff --git a/block-cow.c b/block-cow.c
index 777d8a5..9ad0a32 100644
--- a/block-cow.c
+++ b/block-cow.c
@@ -224,7 +224,6 @@ static int cow_create(const char *filename, int64_t image_sectors,
 
         fd = open(image_filename, O_RDONLY | O_BINARY);
         if (fd < 0) {
-            close(cow_fd);
             goto mtime_fail;
         }
         if (fstat(fd, &st) != 0) {
