From 87d02afc2ca9fed43129cbf5bf8c865c367fa6ff Mon Sep 17 00:00:00 2001
From: Yunlei Ding <yunlei.ding@citrix.com>
Date: Tue, 11 Mar 2014 10:05:15 +0000
Subject: [PATCH 06/11] net: initialize saddr before getsockname() call
To: andrew.cooper3@citrix.com

Signed-off-by: Yunlei Ding <yunlei.ding@citrix.com>
diff --git a/net.c b/net.c
index 6d52817..7f05184 100644
--- a/net.c
+++ b/net.c
@@ -1325,6 +1325,8 @@ static NetSocketState *net_socket_fd_init_dgram(VLANState *vlan,
     socklen_t saddr_len = sizeof(saddr);
     NetSocketState *s;
 
+    memset(&saddr, 0, sizeof(saddr));
+
     /* fd passed: multicast: "learn" dgram_dst address from bound address and save it
      * Because this may be "shared" socket from a "master" process, datagrams would be recv()
      * by ONLY ONE process: we must "clone" this dgram socket --jjo
