From 0772418e7165351b004860af9ff4562444df439e Mon Sep 17 00:00:00 2001
From: Kaifeng Zhu <kaifeng.zhu@citrix.com>
Date: Fri, 7 Mar 2014 16:31:06 +0000
Subject: [PATCH 07/10] net: don't leak an fd after an error
To: andrew.cooper3@citrix.com

fd will be leaked if launch_script failed.

Signed-off-by: Kaifeng Zhu <kaifeng.zhu@citrix.com>
diff --git a/net.c b/net.c
index d47ec2a..a9ba781 100644
--- a/net.c
+++ b/net.c
@@ -1054,8 +1054,10 @@ static int net_tap_init(VLANState *vlan, const char *model,
     if (!setup_script || !strcmp(setup_script, "no"))
         setup_script = "";
     if (setup_script[0] != '\0') {
-	if (launch_script(setup_script, ifname, script_arg, fd))
+	if (launch_script(setup_script, ifname, script_arg, fd)) {
+	    close(fd);
 	    return -1;
+	}
     }
     s = net_tap_fd_init(vlan, model, name, fd);
     if (!s)
