From 98e699ceef6d307ee9e931e68a7120ddac1856a9 Mon Sep 17 00:00:00 2001
From: Yunlei Ding <yunlei.ding@citrix.com>
Date: Tue, 11 Mar 2014 10:12:01 +0000
Subject: [PATCH 08/11] block-nbd: close sock in nbd_open() error path
To: andrew.cooper3@citrix.com

Close sock handle before return.

Signed-off-by: Yunlei Ding <yunlei.ding@citix.com>
Signed-off-by: Yunlei Ding <yunlei.ding@citrix.com>
diff --git a/block-nbd.c b/block-nbd.c
index dc63183..e2c90eb 100644
--- a/block-nbd.c
+++ b/block-nbd.c
@@ -88,7 +88,10 @@ static int nbd_open(BlockDriverState *bs, const char* filename, int flags)
 
     ret = nbd_receive_negotiate(sock, &size, &blocksize);
     if (ret == -1)
+    {
+        close(sock);
         return -errno;
+    }
 
     s->sock = sock;
     s->size = size;
