From 4964976a3fedfb5877be5d86d8dce42819373d7a Mon Sep 17 00:00:00 2001
From: Yunlei Ding <yunlei.ding@citrix.com>
Date: Tue, 11 Mar 2014 10:19:23 +0000
Subject: [PATCH 10/11] block-vvfat: fix memory leak in read_directory()
To: andrew.cooper3@citrix.com

Free allocated memory if create mapping failed.

Signed-off-by: Yunlei Ding <yunlei.ding@citrix.com>
diff --git a/block-vvfat.c b/block-vvfat.c
index 0ab4176..b6b5cc4 100644
--- a/block-vvfat.c
+++ b/block-vvfat.c
@@ -780,6 +780,8 @@ static int read_directory(BDRVVVFATState* s, int mapping_index)
 	    s->current_mapping->read_only =
 		(st.st_mode & (S_IWUSR | S_IWGRP | S_IWOTH)) == 0;
 	}
+        else
+            qemu_free(buffer);
     }
     closedir(dir);
 
