diff --git a/vl.c b/vl.c
index f140c84..804e69a 100644
--- a/vl.c
+++ b/vl.c
@@ -6092,8 +6092,6 @@ int main(int argc, char **argv, char **envp)
         if (devname && strcmp(devname, "none")) {
             char label[32];
             snprintf(label, sizeof(label), "serial%d", i);
-            if (strstart(devname, "vc", 0))
-                qemu_chr_printf(serial_hds[i], "serial%d console\r\n", i);
             xenstore_store_serial_port_info(i, serial_hds[i], devname);
         }
     }
@@ -6103,8 +6101,6 @@ int main(int argc, char **argv, char **envp)
         if (devname && strcmp(devname, "none")) {
             char label[32];
             snprintf(label, sizeof(label), "parallel%d", i);
-            if (strstart(devname, "vc", 0))
-                qemu_chr_printf(parallel_hds[i], "parallel%d console\r\n", i);
         }
     }
 
