diff --git a/xenstore.c b/xenstore.c
index 8320026..c62b77a 100644
--- a/xenstore.c
+++ b/xenstore.c
@@ -330,13 +330,14 @@ static void xenstore_get_backend_path(char **backend, const char *devtype,
     backend_dompath = xs_get_domain_path(xsh, domid_backend);
     if (!backend_dompath) goto out;
     
-    const char *expected_devtypes[4];
+    const char *expected_devtypes[5];
     const char **expected_devtype = expected_devtypes;
 
     *expected_devtype++ = devtype;
     if (!strcmp(devtype, "vbd")) {
 	*expected_devtype++ = "tap";
 	*expected_devtype++ = "qdisk";
+	*expected_devtype++ = "vbd3";
     }
     *expected_devtype = 0;
     assert(expected_devtype <
