diff --git a/hw/vga.c b/hw/vga.c
index a7812bb..a0930f2 100644
--- a/hw/vga.c
+++ b/hw/vga.c
@@ -1246,7 +1246,8 @@ static void vga_draw_text(VGAState *s, int full_update)
     vga_draw_glyph9_func *vga_draw_glyph9;
 
     /* Disable dirty bit tracking */
-    xc_hvm_track_dirty_vram(xc_handle, domid, 0, 0, NULL);
+    if (xen_logdirty_enable)
+        xc_hvm_track_dirty_vram(xc_handle, domid, 0, 0, NULL);
 
     /* compute font data address (in plane 2) */
     v = s->sr[3];
@@ -1707,7 +1708,7 @@ static void vga_draw_graphic(VGAState *s, int full_update)
             /* We will not read that anyway. */
             cpu_physical_memory_set_dirty(s->vram_offset + y);
 
-        {
+        if (xen_logdirty_enable) {
             unsigned long npages = (end - y) / TARGET_PAGE_SIZE;
             const int width = sizeof(unsigned long) * 8;
             unsigned long bitmap[(npages + width - 1) / width];
@@ -1826,7 +1827,8 @@ static void vga_draw_blank(VGAState *s, int full_update)
         return;
 
     /* Disable dirty bit tracking */
-    xc_hvm_track_dirty_vram(xc_handle, domid, 0, 0, NULL);
+    if (xen_logdirty_enable)
+        xc_hvm_track_dirty_vram(xc_handle, domid, 0, 0, NULL);
 
     s->rgb_to_pixel =
         rgb_to_pixel_dup_table[get_depth_index(s->ds)];
