diff --git a/net.h b/net.h
index 1071fdc..d2e5220 100644
--- a/net.h
+++ b/net.h
@@ -108,8 +108,8 @@ void net_host_device_add(const char *device, const char *opts);
 void net_host_device_remove(int vlan_id, const char *device);
 
 #ifndef DEFAULT_NETWORK_SCRIPT
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
+#define DEFAULT_NETWORK_SCRIPT ""
+#define DEFAULT_NETWORK_DOWN_SCRIPT ""
 #endif
 #ifdef __sun__
 #define SMBD_COMMAND "/usr/sfw/sbin/smbd"
diff --git a/xen-config-host.h b/xen-config-host.h
index 3a02021..d5fa42e 100644
--- a/xen-config-host.h
+++ b/xen-config-host.h
@@ -48,4 +48,7 @@ extern unsigned int xen_logdirty_enable;
 #undef HAVE_IOVEC
 #endif
 
+#undef DEFAULT_NETWORK_SCRIPT
+#undef DEFAULT_NETWORK_DOWN_SCRIPT
+
 #endif /*XEN_CONFIG_HOST_H*/
diff --git a/xen-setup b/xen-setup
index c3af79b..39d8070 100755
--- a/xen-setup
+++ b/xen-setup
@@ -29,8 +29,6 @@ ln -sf ${QEMU_ROOT:-..}/xen-config.mak $target/config.mak
 cat ${QEMU_ROOT:-.}/xen-config-host.mak >>config-host.mak
 
 sed -e 's,qemu,xen/qemu,' config-host.h >config-host.h.new
-echo "#define DEFAULT_NETWORK_SCRIPT      \"${XEN_SCRIPT_DIR}/qemu-ifup\"" >>config-host.h.new
-echo "#define DEFAULT_NETWORK_DOWN_SCRIPT \"${XEN_SCRIPT_DIR}/qemu-ifdown\"" >>config-host.h.new
 echo '#include "xen-config-host.h"' >>config-host.h.new
 
 if test -f config-host.h~; then mv config-host.h~ config-host.h; fi
