diff --git a/hw/cirrus_vga.c b/hw/cirrus_vga.c
index 2f7a5c7..9b72157 100644
--- a/hw/cirrus_vga.c
+++ b/hw/cirrus_vga.c
@@ -854,13 +854,9 @@ static void cirrus_bitblt_reset(CirrusVGAState * s)
 
     s->gr[0x31] &=
 	~(CIRRUS_BLT_START | CIRRUS_BLT_BUSY | CIRRUS_BLT_FIFOUSED);
-    need_update = s->cirrus_srcptr != &s->cirrus_bltbuf[0]
-        || s->cirrus_srcptr_end != &s->cirrus_bltbuf[0];
     s->cirrus_srcptr = &s->cirrus_bltbuf[0];
     s->cirrus_srcptr_end = &s->cirrus_bltbuf[0];
     s->cirrus_srccounter = 0;
-    if (!need_update)
-        return;
     cirrus_update_memory_access(s);
 }
 
