diff --git a/Makefile.target b/Makefile.target
index 8c07dd9..2cf50b4 100644
--- a/Makefile.target
+++ b/Makefile.target
@@ -586,6 +586,7 @@ ifdef CONFIG_AUDIO
 OBJS+= $(SOUND_HW) pcspk.o
 CPPFLAGS += -DHAS_AUDIO -DHAS_AUDIO_CHOICE
 endif
+CPPFLAGS += -I/usr/include/libdrm
 OBJS+= ide.o pckbd.o ps2.o vga.o vgpu_fb.o dma.o
 OBJS+= fdc.o mc146818rtc.o serial.o i8259.o i8254.o pc.o
 OBJS+= cirrus_vga.o apic.o parallel.o acpi.o piix_pci.o
