# HG changeset patch
# Parent 3940be5a37e8d36567384e9eb19a40446edbf4d1

diff --git a/i386-dm/helper2.c b/i386-dm/helper2.c
index fa86daa..2f3bb54 100644
--- a/i386-dm/helper2.c
+++ b/i386-dm/helper2.c
@@ -638,6 +638,11 @@ int main_loop(void)
             FD_SET(xenstore_fd(), &fds);
             if (select(xenstore_fd() + 1, &fds, NULL, NULL, NULL) > 0)
                 xenstore_process_event(NULL);
+            if (qemu_exit_requested()) {
+                /* just exit, we already saved all disks */
+                destroy_hvm_domain();
+                exit(0);
+            }
         }
 
         xenstore_record_dm_state("running");
