diff --git a/hw/intel-tools/intel_batchbuffer.c b/hw/intel-tools/intel_batchbuffer.c
index 789a859..f61736e 100644
--- a/hw/intel-tools/intel_batchbuffer.c
+++ b/hw/intel-tools/intel_batchbuffer.c
@@ -240,6 +240,14 @@ intel_batchbuffer_emit_reloc(struct intel_batchbuffer *batch,
 	igt_assert(ret == 0);
 }
 
+#define BCS_SWCTRL                      0x22200
+#define BCS_SWCTRL_SRC_Y               (1 << 0)
+#define BCS_SWCTRL_DST_Y               (1 << 1)
+
+#define CMD_MI				(0x0 << 29)
+
+#define MI_FLUSH_DW			(CMD_MI | (0x26 << 23) | 2)
+
 /**
  * intel_blt_copy:
  * @batch: batchbuffer object
@@ -315,7 +323,28 @@ intel_blt_copy(struct intel_batchbuffer *batch,
 		igt_fail(IGT_EXIT_FAILURE);
 	}
 
-	BLIT_COPY_BATCH_START(cmd_bits);
+    if (gen >= 9 && ((dst_tiling == I915_TILING_Y) || (src_tiling == I915_TILING_Y))) {
+	    BEGIN_BATCH(22, 2); \
+        //  Switch blitter to y-tiling mode 
+        OUT_BATCH(MI_FLUSH_DW);
+        OUT_BATCH(0);
+        OUT_BATCH(0);
+        OUT_BATCH(0);
+
+        OUT_BATCH(MI_LOAD_REGISTER_IMM | (3 - 2));
+        OUT_BATCH(BCS_SWCTRL);
+        OUT_BATCH((BCS_SWCTRL_DST_Y | BCS_SWCTRL_SRC_Y) << 16 |
+                 ((dst_tiling == I915_TILING_Y) ? BCS_SWCTRL_DST_Y : 0) |
+                 ((src_tiling == I915_TILING_Y) ? BCS_SWCTRL_SRC_Y : 0));
+	    OUT_BATCH(XY_SRC_COPY_BLT_CMD | \
+		  XY_SRC_COPY_BLT_WRITE_ALPHA | \
+		  XY_SRC_COPY_BLT_WRITE_RGB | \
+		  cmd_bits | \
+		  (6 + 2*(batch->gen >= 8))); \
+    } else {
+	    BLIT_COPY_BATCH_START(cmd_bits);
+    }
+
 	OUT_BATCH((br13_bits) |
 		  (0xcc << 16) | /* copy ROP */
 		  dst_pitch);
@@ -325,6 +354,19 @@ intel_blt_copy(struct intel_batchbuffer *batch,
 	OUT_BATCH((src_y1 << 16) | src_x1); /* src x1,y1 */
 	OUT_BATCH(src_pitch);
 	OUT_RELOC_FENCED(src_bo, I915_GEM_DOMAIN_RENDER, 0, 0);
+
+    if (gen >= 9 && ((dst_tiling == I915_TILING_Y) || (src_tiling == I915_TILING_Y))) {
+       //  Switch blitter back to x-tiling mode 
+       OUT_BATCH(MI_FLUSH_DW);
+       OUT_BATCH(0);
+       OUT_BATCH(0);
+       OUT_BATCH(0);
+
+       OUT_BATCH(MI_LOAD_REGISTER_IMM | (3 - 2));
+       OUT_BATCH(BCS_SWCTRL);
+       OUT_BATCH((BCS_SWCTRL_DST_Y | BCS_SWCTRL_SRC_Y) << 16 );
+    }
+
 	ADVANCE_BATCH();
 
 #define CMD_POLY_STIPPLE_OFFSET       0x7906
diff --git a/hw/intel-tools/intel_batchbuffer.h b/hw/intel-tools/intel_batchbuffer.h
index 7e957c1..7f8876f 100644
--- a/hw/intel-tools/intel_batchbuffer.h
+++ b/hw/intel-tools/intel_batchbuffer.h
@@ -173,10 +173,28 @@ intel_batchbuffer_require_space(struct intel_batchbuffer *batch,
 		  (4 + (batch->gen >= 8))); \
 } while(0)
 
+/*
+ * Yf/Ys tiling
+ *
+ * Tiling mode in the I915_TILING_... namespace for new tiling modes which are
+ * defined in the kernel. (They are not fenceable so the kernel does not need
+ * to know about them.)
+ *
+ * They are to be used the the blitting routines below.
+ */
+#define I915_TILING_Yf	3
+#define I915_TILING_Ys	4
+
 void
 intel_blt_copy(struct intel_batchbuffer *batch,
 	      drm_intel_bo *src_bo, int src_x1, int src_y1, int src_pitch,
 	      drm_intel_bo *dst_bo, int dst_x1, int dst_y1, int dst_pitch,
 	      int width, int height, int bpp);
 
+void igt_blitter_fast_copy(struct intel_batchbuffer *batch,
+			   drm_intel_bo *src_bo, uint32_t src_stride, uint32_t src_tiling, unsigned src_x, unsigned src_y,
+			   unsigned width, unsigned height,
+			   drm_intel_bo *dst_bo, uint32_t dst_stride, uint32_t dst_tiling, unsigned dst_x, unsigned dst_y);
+
+
 #endif
diff --git a/hw/intel-tools/intel_chipset.h b/hw/intel-tools/intel_chipset.h
index 7f611ed..4b40612 100644
--- a/hw/intel-tools/intel_chipset.h
+++ b/hw/intel-tools/intel_chipset.h
@@ -198,17 +198,46 @@ void intel_check_pch(void);
 #define PCI_CHIP_SKYLAKE_ULX_GT2	0x191E
 #define PCI_CHIP_SKYLAKE_DT_GT2		0x1912
 #define PCI_CHIP_SKYLAKE_DT_GT1		0x1902
+#define PCI_CHIP_SKYLAKE_DT_GT4		0x1932
 #define PCI_CHIP_SKYLAKE_HALO_GT2	0x191B
 #define PCI_CHIP_SKYLAKE_HALO_GT3	0x192B
 #define PCI_CHIP_SKYLAKE_HALO_GT1 	0x190B
+#define PCI_CHIP_SKYLAKE_HALO_GT4 	0x193B
 #define PCI_CHIP_SKYLAKE_SRV_GT2	0x191A
 #define PCI_CHIP_SKYLAKE_SRV_GT3	0x192A
 #define PCI_CHIP_SKYLAKE_SRV_GT1	0x190A
+#define PCI_CHIP_SKYLAKE_SRV_GT4	0x193A
 #define PCI_CHIP_SKYLAKE_WKS_GT2 	0x191D
+#define PCI_CHIP_SKYLAKE_WKS_GT4 	0x193D
+
+#define PCI_CHIP_KABYLAKE_ULT_GT2      0x5916
+#define PCI_CHIP_KABYLAKE_ULT_GT1_5    0x5913
+#define PCI_CHIP_KABYLAKE_ULT_GT1      0x5906
+#define PCI_CHIP_KABYLAKE_ULT_GT3      0x5926
+#define PCI_CHIP_KABYLAKE_ULT_GT2F     0x5921
+#define PCI_CHIP_KABYLAKE_ULX_GT1_5    0x5915
+#define PCI_CHIP_KABYLAKE_ULX_GT1      0x590E
+#define PCI_CHIP_KABYLAKE_ULX_GT2      0x591E
+#define PCI_CHIP_KABYLAKE_DT_GT2       0x5912
+#define PCI_CHIP_KABYLAKE_DT_GT1_5     0x5917
+#define PCI_CHIP_KABYLAKE_DT_GT1       0x5902
+#define PCI_CHIP_KABYLAKE_DT_GT4       0x5932
+#define PCI_CHIP_KABYLAKE_HALO_GT2     0x591B
+#define PCI_CHIP_KABYLAKE_HALO_GT3     0x592B
+#define PCI_CHIP_KABYLAKE_HALO_GT1     0x590B
+#define PCI_CHIP_KABYLAKE_HALO_GT4     0x593B
+#define PCI_CHIP_KABYLAKE_SRV_GT2      0x591A
+#define PCI_CHIP_KABYLAKE_SRV_GT3      0x592A
+#define PCI_CHIP_KABYLAKE_SRV_GT4      0x593A
+#define PCI_CHIP_KABYLAKE_SRV_GT1      0x590A
+#define PCI_CHIP_KABYLAKE_WKS_GT2      0x591D
+#define PCI_CHIP_KABYLAKE_WKS_GT4      0x593D
 
 #define PCI_CHIP_BROXTON_0		0x0A84
 #define PCI_CHIP_BROXTON_1		0x1A84
 #define PCI_CHIP_BROXTON_2		0x5A84
+#define PCI_CHIP_BROXTON_3		0x1A85
+#define PCI_CHIP_BROXTON_4		0x5A85
 
 #endif /* __GTK_DOC_IGNORE__ */
 
@@ -390,15 +419,56 @@ void intel_check_pch(void);
 				 (devid) == PCI_CHIP_SKYLAKE_HALO_GT3	|| \
 				 (devid) == PCI_CHIP_SKYLAKE_SRV_GT3)
 
+#define IS_SKL_GT4(devid)	((devid) == PCI_CHIP_SKYLAKE_DT_GT4	|| \
+				 (devid) == PCI_CHIP_SKYLAKE_HALO_GT4	|| \
+				 (devid) == PCI_CHIP_SKYLAKE_WKS_GT4	|| \
+				 (devid) == PCI_CHIP_SKYLAKE_SRV_GT4)
+
+#define IS_KBL_GT1(devid)	((devid) == PCI_CHIP_KABYLAKE_ULT_GT1_5|| \
+				 (devid) == PCI_CHIP_KABYLAKE_ULX_GT1_5|| \
+				 (devid) == PCI_CHIP_KABYLAKE_DT_GT1_5|| \
+				 (devid) == PCI_CHIP_KABYLAKE_ULT_GT1|| \
+				 (devid) == PCI_CHIP_KABYLAKE_ULX_GT1|| \
+				 (devid) == PCI_CHIP_KABYLAKE_DT_GT1||	\
+				 (devid) == PCI_CHIP_KABYLAKE_HALO_GT1|| \
+				 (devid) == PCI_CHIP_KABYLAKE_SRV_GT1)
+
+#define IS_KBL_GT2(devid)	((devid) == PCI_CHIP_KABYLAKE_ULT_GT2|| \
+				 (devid) == PCI_CHIP_KABYLAKE_ULT_GT2F|| \
+				 (devid) == PCI_CHIP_KABYLAKE_ULX_GT2|| \
+				 (devid) == PCI_CHIP_KABYLAKE_DT_GT2||	\
+				 (devid) == PCI_CHIP_KABYLAKE_HALO_GT2|| \
+				 (devid) == PCI_CHIP_KABYLAKE_SRV_GT2|| \
+				 (devid) == PCI_CHIP_KABYLAKE_WKS_GT2)
+
+#define IS_KBL_GT3(devid)	((devid) == PCI_CHIP_KABYLAKE_ULT_GT3|| \
+				 (devid) == PCI_CHIP_KABYLAKE_HALO_GT3|| \
+				 (devid) == PCI_CHIP_KABYLAKE_SRV_GT3)
+
+#define IS_KBL_GT4(devid)	((devid) == PCI_CHIP_KABYLAKE_DT_GT4|| \
+				 (devid) == PCI_CHIP_KABYLAKE_HALO_GT4|| \
+				 (devid) == PCI_CHIP_KABYLAKE_SRV_GT4|| \
+				 (devid) == PCI_CHIP_KABYLAKE_WKS_GT4)
+
+#define IS_KABYLAKE(devid)	(IS_KBL_GT1(devid) || \
+ 				 IS_KBL_GT2(devid) || \
+ 				 IS_KBL_GT3(devid) || \
+ 				 IS_KBL_GT4(devid))
+
 #define IS_SKYLAKE(devid)	(IS_SKL_GT1(devid) || \
 				 IS_SKL_GT2(devid) || \
-				 IS_SKL_GT3(devid))
+				 IS_SKL_GT3(devid) || \
+				 IS_SKL_GT4(devid))
 
 #define IS_BROXTON(devid)	((devid) == PCI_CHIP_BROXTON_0 || \
 				 (devid) == PCI_CHIP_BROXTON_1 || \
-				 (devid) == PCI_CHIP_BROXTON_2)
+				 (devid) == PCI_CHIP_BROXTON_2 || \
+				 (devid) == PCI_CHIP_BROXTON_3 || \
+				 (devid) == PCI_CHIP_BROXTON_4)
 
-#define IS_GEN9(devid)		(IS_SKYLAKE(devid) || IS_BROXTON(devid))
+#define IS_GEN9(devid)		(IS_KABYLAKE(devid) || \
+				 IS_SKYLAKE(devid) || \
+				 IS_BROXTON(devid))
 
 #define IS_965(devid)		(IS_GEN4(devid) || \
 				 IS_GEN5(devid) || \
diff --git a/hw/intel-tools/intel_reg.h b/hw/intel-tools/intel_reg.h
index 18a68e9..0ffa803 100644
--- a/hw/intel-tools/intel_reg.h
+++ b/hw/intel-tools/intel_reg.h
@@ -36,8 +36,8 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  *   Eric Anholt <eric@anholt.net>
  *
  *   based on the i740 driver by
- *        Kevin E. Martin <kevin@precisioninsight.com>
- *
+ *        Kevin E. Martin <kevin@precisioninsight.com> 
+ *   
  *
  */
 
@@ -68,14 +68,14 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define EXT_HORIZ_TOTAL      0x35 /* p261 */
 #define EXT_HORIZ_BLANK      0x39 /* p261 */
 #define EXT_START_ADDR       0x40 /* p262 */
-#define EXT_START_ADDR_ENABLE    0x80
+#define EXT_START_ADDR_ENABLE    0x80 
 #define EXT_OFFSET           0x41 /* p263 */
 #define EXT_START_ADDR_HI    0x42 /* p263 */
 #define INTERLACE_CNTL       0x70 /* p264 */
-#define INTERLACE_ENABLE         0x80
-#define INTERLACE_DISABLE        0x00
+#define INTERLACE_ENABLE         0x80 
+#define INTERLACE_DISABLE        0x00 
 
-/* Miscellaneous Output Register
+/* Miscellaneous Output Register 
  */
 #define MSR_R          0x3CC	/* p207 */
 #define MSR_W          0x3C2	/* p207 */
@@ -128,7 +128,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define C1DRA01			0x10608
 #define C1DRA23			0x1060a
 
-/* p375.
+/* p375. 
  */
 #define DISPLAY_CNTL       0x70008
 #define VGA_WRAP_MODE          0x02
@@ -348,11 +348,11 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define VCO_LOOP_DIV_BY_16M    0x04
 
 
-/* Instruction Parser Mode Register
+/* Instruction Parser Mode Register 
  *    - p281
  *    - 2 new bits.
  */
-#define INST_PM                  0x20c0
+#define INST_PM                  0x20c0	
 #define AGP_SYNC_PACKET_FLUSH_ENABLE 0x20 /* reserved */
 #define SYNC_PACKET_FLUSH_ENABLE     0x10
 #define TWO_D_INST_DISABLE           0x08
@@ -381,7 +381,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define GEN6_INSTDONE_1		0x206c
 #define INST_PS_I965                0x2070
 
-/* Current active ring head address:
+/* Current active ring head address: 
  */
 #define ACTHD_I965                 0x2074
 #define ACTHD			   0x20C8
@@ -466,7 +466,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 # define ERR_INSTRUCTION_ERROR			(1 << 0)
 
 
-/* Interrupt Control Registers
+/* Interrupt Control Registers 
  *   - new bits for i810
  *   - new register hwstam (mask)
  */
@@ -502,7 +502,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 				  HOST_PORT_EVENT |	\
 				  CAPTURE_EVENT )
 
-/* FIFO Watermark and Burst Length Control Register
+/* FIFO Watermark and Burst Length Control Register 
  *
  * - different offset and contents on i810 (p299) (fewer bits per field)
  * - some overlay fields added
@@ -663,7 +663,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define I830_RING_START_MASK	0xFFFFF000
 
 #define RING_LEN       0x0C
-#define RING_NR_PAGES       0x001FF000
+#define RING_NR_PAGES       0x001FF000 
 #define I830_RING_NR_PAGES	0x001FF000
 #define RING_REPORT_MASK    0x00000006
 #define RING_REPORT_64K     0x00000002
@@ -749,7 +749,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 
 /* Registers in the i810 host-pci bridge pci config space which affect
- * the i810 graphics operations.
+ * the i810 graphics operations.  
  */
 #define SMRAM_MISCC         0x70
 #define GMS                    0x000000c0
@@ -757,11 +757,11 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define GMS_ENABLE_BARE        0x00000040
 #define GMS_ENABLE_512K        0x00000080
 #define GMS_ENABLE_1M          0x000000c0
-#define USMM                   0x00000030
+#define USMM                   0x00000030 
 #define USMM_DISABLE           0x00000000
 #define USMM_TSEG_ZERO         0x00000010
 #define USMM_TSEG_512K         0x00000020
-#define USMM_TSEG_1M           0x00000030
+#define USMM_TSEG_1M           0x00000030  
 #define GFX_MEM_WIN_SIZE       0x00010000
 #define GFX_MEM_WIN_32M        0x00010000
 #define GFX_MEM_WIN_64M        0x00000000
@@ -1036,7 +1036,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 # define DPLL_MD_UDI_MULTIPLIER_MASK		0x00003f00
 # define DPLL_MD_UDI_MULTIPLIER_SHIFT		8
-/* SDVO/UDI pixel multiplier for VGA, same as DPLL_MD_UDI_MULTIPLIER_MASK.
+/* SDVO/UDI pixel multiplier for VGA, same as DPLL_MD_UDI_MULTIPLIER_MASK. 
  * This best be set to the default value (3) or the CRT won't work. No,
  * I don't entirely understand what this does...
  */
@@ -2213,7 +2213,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 # define SVBLANK_INT_STATUS	(1 << 2)
 # define VBLANK_INT_STATUS	(1 << 1)
 # define OREG_UPDATE_STATUS	(1 << 0)
-
+				 
 #define FW_BLC		0x020d8
 #define FW_BLC2		0x020dc
 #define FW_BLC_SELF	0x020e0 /* 915+ only */
@@ -2228,7 +2228,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #define DSPFW3			0x7003c
 /*
  * The two pipe frame counter registers are not synchronized, so
- * reading a stable value is somewhat tricky. The following code
+ * reading a stable value is somewhat tricky. The following code 
  * should work:
  *
  *  do {
