diff --git a/mk/xen-device-model.spec.in b/mk/xen-device-model.spec.in
index c86cbe8..cec1236 100644
--- a/mk/xen-device-model.spec.in
+++ b/mk/xen-device-model.spec.in
@@ -8,7 +8,6 @@ Source0: xen-device-model-%{version}.tar.bz2
 Patch0: xen-device-model-development.patch
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
 BuildRequires: zlib-devel, xen-devel, xen-dom0-devel, pciutils-devel, check-devel
-BuildRequires: xenserver-libdrm-devel
 Requires(pre): shadow-utils
 Requires: libdrm
 Provides: qemu-xen(syslog) = 1
