diff --git a/cpu-all.h b/cpu-all.h
index a581c54..7c704cd 100644
--- a/cpu-all.h
+++ b/cpu-all.h
@@ -865,6 +865,7 @@ extern ram_addr_t ram_size;
 extern int vgt_low_gm_sz;
 extern int vgt_high_gm_sz;
 extern int vgt_fence_sz;
+extern int vgt_cap;
 extern const char *vgt_monitor_config_file;
 
 /* physical memory access */
diff --git a/hw/vga-xengt.c b/hw/vga-xengt.c
index a4f2196..7815a26 100644
--- a/hw/vga-xengt.c
+++ b/hw/vga-xengt.c
@@ -90,6 +90,7 @@ int vgt_low_gm_sz = 128; /* in MB */
 int vgt_high_gm_sz = 448; /* in MB */
 int vgt_fence_sz = 4;
 int vgt_primary = 1; /* -1 means "not specified */
+int vgt_cap = 0;
 const char *vgt_monitor_config_file = "/etc/gvt-g-monitor.conf";
 
 int vgt_legacy_vga_ram_size;
@@ -867,11 +868,13 @@ static void create_vgt_instance(void)
     int err = 0;
 
     qemu_log("vGT: %s: domid=%d, low_gm_sz=%dMB, high_gm_sz=%dMB, "
-        "fence_sz=%d, vga_sz=%dMB, vgt_primary=%d\n", __func__, xen_domid,
-        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, 8, vgt_primary);
+             "fence_sz=%d, vga_sz=%dMB, vgt_primary=%d vgt_cap=%d\n",
+             __func__, xen_domid, vgt_low_gm_sz, vgt_high_gm_sz,
+             vgt_fence_sz, 8, vgt_primary, vgt_cap);
     if (vgt_low_gm_sz <= 0 || vgt_high_gm_sz <=0 ||
-            vgt_primary < -1 || vgt_primary > 1 ||
-            vgt_fence_sz <=0) {
+        vgt_cap < 0 || vgt_cap > 100 ||
+        vgt_primary < -1 || vgt_primary > 1 ||
+        vgt_fence_sz <=0) {
         qemu_log("vGT: %s failed: invalid parameters!\n", __func__);
         abort();
     }
@@ -885,8 +888,10 @@ static void create_vgt_instance(void)
      * driver to create a vgt instanc for Domain domid with the required
      * parameters. NOTE: aperture_size and gm_size are in MB.
      */
-    if (!err && fprintf(vgt_file, "%d,%u,%u,%u,%d\n", xen_domid,
-        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary) < 0) {
+    if (!err && 
+        fprintf(vgt_file, "%d,%u,%u,%u,%d,%d\n", xen_domid,
+                vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary,
+                vgt_cap) < 0) {
         err = errno;
     }
 
diff --git a/vl.c b/vl.c
index 82bf379..89fcf11 100644
--- a/vl.c
+++ b/vl.c
@@ -3884,6 +3884,7 @@ enum {
     QEMU_OPTION_vgt_low_gm_sz,
     QEMU_OPTION_vgt_high_gm_sz,
     QEMU_OPTION_vgt_fence_sz,
+    QEMU_OPTION_vgt_cap,
     QEMU_OPTION_vgt_monitor_config_file,
 
     /* Debug/Expert options: */
@@ -4077,6 +4078,7 @@ static const QEMUOption qemu_options[] = {
     { "vgt_low_gm_sz", HAS_ARG, QEMU_OPTION_vgt_low_gm_sz },
     { "vgt_high_gm_sz", HAS_ARG, QEMU_OPTION_vgt_high_gm_sz },
     { "vgt_fence_sz", HAS_ARG, QEMU_OPTION_vgt_fence_sz },
+    { "vgt_cap", HAS_ARG, QEMU_OPTION_vgt_cap },
     { "vgt_monitor_config_file", HAS_ARG, QEMU_OPTION_vgt_monitor_config_file },
 
 #if defined(TARGET_ARM)
@@ -5074,6 +5076,12 @@ int main(int argc, char **argv, char **envp)
                     vgt_fence_sz = strtol(optarg,&ptr,10);
                 }
                 break;
+            case QEMU_OPTION_vgt_cap:
+                {
+                    char *ptr;
+                    vgt_cap = strtol(optarg,&ptr,10);
+                }
+                break;
             case QEMU_OPTION_vgt_monitor_config_file:
                 {
                     vgt_monitor_config_file = optarg;
