# HG changeset patch
# Parent d55764563b44c59ab220469ca7c3befe7f38e14c

diff --git a/vnc.c b/vnc.c
index ed6de92..91f6ec6 100644
--- a/vnc.c
+++ b/vnc.c
@@ -88,7 +88,7 @@ typedef void VncSendHextileTile(VncState *vs,
                                 void *last_fg,
                                 int *has_bg, int *has_fg);
 
-#define VNC_MAX_WIDTH 2048
+#define VNC_MAX_WIDTH 2560
 #define VNC_MAX_HEIGHT 2048
 #define VNC_DIRTY_WORDS (VNC_MAX_WIDTH / (16 * 32))
 
