diff --git a/xenstore.c b/xenstore.c
index 909e07e..cb96371 100644
--- a/xenstore.c
+++ b/xenstore.c
@@ -1897,9 +1897,9 @@ int store_dev_info(const char *devName, int domid,
     fprintf(logfile, "can't store dev %s name for domid %d in %s from a stub domain\n", devName, domid, storeString);
     return ENOSYS;
 #else
-    xc_interface *xc_handle;
-    struct xs_handle *xs;
-    char *path;
+    xc_interface *xc_handle = NULL;
+    struct xs_handle *xs = NULL;
+    char *path = NULL;
     char *newpath;
     char *pts;
     char namebuf[128];
@@ -1917,31 +1917,31 @@ int store_dev_info(const char *devName, int domid,
     }
     if (memcmp(namebuf, "pty ", 4)) return 0;
     pts = namebuf + 4;
+    ret = -1;
 
     /* We now have everything we need to set the xenstore entry. */
     xs = xs_daemon_open();
     if (xs == NULL) {
         fprintf(logfile, "Could not contact XenStore\n");
-        return -1;
+        goto out;
     }
 
     xc_handle = xc_interface_open(0,0,0);
     if (xc_handle == NULL) {
         fprintf(logfile, "xc_interface_open() error\n");
-        return -1;
+        goto out;
     }
 
     path = xs_get_domain_path(xs, domid);
     if (path == NULL) {
         fprintf(logfile, "xs_get_domain_path() error\n");
-        return -1;
+        goto out;
     }
     newpath = realloc(path, (strlen(path) + strlen(storeString) +
                              strlen("/tty") + 1));
     if (newpath == NULL) {
-        free(path); /* realloc errors leave old block */
         fprintf(logfile, "realloc error\n");
-        return -1;
+        goto out;
     }
     path = newpath;
 
@@ -1949,14 +1949,19 @@ int store_dev_info(const char *devName, int domid,
     strcat(path, "/tty");
     if (!xs_write(xs, XBT_NULL, path, pts, strlen(pts))) {
         fprintf(logfile, "xs_write for '%s' fail", storeString);
-        return -1;
+        goto out;
     }
 
+    ret = 0;
+ out:
+
     free(path);
-    xs_daemon_close(xs);
-    xc_interface_close(xc_handle);
+    if (xs)
+        xs_daemon_close(xs);
+    if (xc_handle)
+        xc_interface_close(xc_handle);
 
-    return 0;
+    return ret;
 #endif
 }
 
