diff --git a/mk/xen-device-model.spec.in b/mk/xen-device-model.spec.in
index debc12f..cec1236 100644
--- a/mk/xen-device-model.spec.in
+++ b/mk/xen-device-model.spec.in
@@ -9,6 +9,7 @@ Patch0: xen-device-model-development.patch
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
 BuildRequires: zlib-devel, xen-devel, xen-dom0-devel, pciutils-devel, check-devel
 Requires(pre): shadow-utils
+Requires: libdrm
 Provides: qemu-xen(syslog) = 1
 
 %description
