diff --git a/mk/xen-device-model.spec.in b/mk/xen-device-model.spec.in
index d1e1482..69d216c 100644
--- a/mk/xen-device-model.spec.in
+++ b/mk/xen-device-model.spec.in
@@ -7,7 +7,8 @@ Group: System/Hypervisor
 Source0: xen-device-model-%{version}.tar.bz2
 Patch0: xen-device-model-development.patch
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
-BuildRequires: zlib-devel, xen-devel, xen-dom0-devel, pciutils-devel, check-devel
+BuildRequires: zlib-devel, xen-devel, xen-dom0-devel, pciutils-devel, check-devel, libdrm-devel
+Requires: libdrm
 Provides: qemu-xen(syslog) = 1
 
 %description
