diff --git a/Rules.mk b/Rules.mk
index afd2295..1ae138b 100644
--- a/Rules.mk
+++ b/Rules.mk
@@ -25,3 +25,15 @@ CFLAGS += -Wstrict-prototypes
 CFLAGS += -Wno-unused-value
 CFLAGS += -m64
 CFLAGS += -Wdeclaration-after-statement
+
+# XenServer hacks
+# Override target output
+LIBEXEC := ${prefix}/libexec/xen/bin
+LIBEXEC_BIN := $(LIBEXEC)
+XEN_SCRIPT_DIR := /etc/xen/scripts
+
+# Don't build img utilities
+TOOLS :=
+
+# Don't link against blktap1
+CONFIG_BLKTAP1=n
diff --git a/xen-setup b/xen-setup
index c3af79b..2da3085 100755
--- a/xen-setup
+++ b/xen-setup
@@ -29,8 +29,8 @@ ln -sf ${QEMU_ROOT:-..}/xen-config.mak $target/config.mak
 cat ${QEMU_ROOT:-.}/xen-config-host.mak >>config-host.mak
 
 sed -e 's,qemu,xen/qemu,' config-host.h >config-host.h.new
-echo "#define DEFAULT_NETWORK_SCRIPT      \"${XEN_SCRIPT_DIR}/qemu-ifup\"" >>config-host.h.new
-echo "#define DEFAULT_NETWORK_DOWN_SCRIPT \"${XEN_SCRIPT_DIR}/qemu-ifdown\"" >>config-host.h.new
+echo "#define DEFAULT_NETWORK_SCRIPT      \"\"" >>config-host.h.new
+echo "#define DEFAULT_NETWORK_DOWN_SCRIPT \"\"" >>config-host.h.new
 echo '#include "xen-config-host.h"' >>config-host.h.new
 
 if test -f config-host.h~; then mv config-host.h~ config-host.h; fi
