diff --git a/Rules.mk b/Rules.mk
index afd2295..dfe271d 100644
--- a/Rules.mk
+++ b/Rules.mk
@@ -25,3 +25,14 @@ CFLAGS += -Wstrict-prototypes
 CFLAGS += -Wno-unused-value
 CFLAGS += -m64
 CFLAGS += -Wdeclaration-after-statement
+
+# XenServer hacks
+# Override target output
+LIBEXEC := ${prefix}/lib64/xen/bin
+XEN_SCRIPT_DIR := /etc/xen/scripts
+
+# Don't build img utilities
+TOOLS :=
+
+# Don't link against blktap1
+CONFIG_BLKTAP1=n
