diff --git a/sysemu.h b/sysemu.h
index fcc75b1..6070dfe 100644
--- a/sysemu.h
+++ b/sysemu.h
@@ -28,9 +28,11 @@ void cpu_disable_ticks(void);
 void qemu_system_reset_request(void);
 void qemu_system_shutdown_request(void);
 void qemu_system_powerdown_request(void);
+void qemu_system_exit_request(void);
 int qemu_shutdown_requested(void);
 int qemu_reset_requested(void);
 int qemu_powerdown_requested(void);
+int qemu_exit_requested(void);
 #if !defined(TARGET_SPARC) && !defined(TARGET_I386)
 // Please implement a power failure function to signal the OS
 #define qemu_system_powerdown() do{}while(0)
diff --git a/vl.c b/vl.c
index 4b2e5fe..e70bc7b 100644
--- a/vl.c
+++ b/vl.c
@@ -3632,6 +3632,7 @@ static QEMUResetEntry *first_reset_entry;
 static int reset_requested;
 static int shutdown_requested;
 static int powerdown_requested;
+static int exit_requested;
 
 int qemu_shutdown_requested(void)
 {
@@ -3654,6 +3655,13 @@ int qemu_powerdown_requested(void)
     return r;
 }
 
+int qemu_exit_requested(void)
+{
+    int r = exit_requested;
+    exit_requested = 0;
+    return r;
+}
+
 void qemu_register_reset(QEMUResetHandler *func, void *opaque)
 {
     QEMUResetEntry **pre, *re;
@@ -3707,6 +3715,11 @@ void qemu_system_powerdown_request(void)
         cpu_interrupt(cpu_single_env, CPU_INTERRUPT_EXIT);
 }
 
+void qemu_system_exit_request(void)
+{
+    exit_requested = 1;
+}
+
 #ifdef _WIN32
 static void host_main_loop_wait(int *timeout)
 {
@@ -3834,6 +3847,12 @@ void main_loop_wait(int timeout)
     }
 #endif
 
+    if (qemu_exit_requested()) {
+        /* Flush the cache and cancel outstanding aio requests */
+        bdrv_flush_all();
+        exit(0);
+    }
+
     /* vm time timers */
     if (vm_running && likely(!(cur_cpu->singlestep_enabled & SSTEP_NOTIMER)))
         qemu_run_timers(&active_timers[QEMU_TIMER_VIRTUAL],
diff --git a/xenstore.c b/xenstore.c
index 56bec53..d0ce0cf 100644
--- a/xenstore.c
+++ b/xenstore.c
@@ -564,6 +564,7 @@ void xenstore_parse_domain_config(int hvm_domid)
     } else {
         guest_path = strdup(danger_path);
     }
+    xs_watch(xsh, "@releaseDomain", "releaseDomain");
 
     if (pasprintf(&buf, "%s/keymap", guest_path) != -1)
         xs_watch(xsh, buf, KEYMAP_TOKEN); // Ignore failure -- we can muddle on.i
@@ -1453,6 +1454,25 @@ void xenstore_process_event(void *opaque)
         goto out;
     }
 
+    if (!strcmp(vec[XS_WATCH_TOKEN], "releaseDomain")) {
+        xc_dominfo_t dominfo;
+        int rc = xc_domain_getinfo(xc_handle, domid, 1, &dominfo);
+
+        if ( rc < 0 ) {
+            fprintf(stderr, "xc_domain_getinfo() failed (%d) %s\n",
+                    errno, strerror(errno));
+            exit(1);
+        } else if ( rc != 1 || dominfo.domid != domid ) {
+            fprintf(stderr, "No domain with domid %d\n", domid);
+            exit(1);
+        } else if ( dominfo.dying )
+            qemu_system_exit_request();
+        else
+            fprintf(stderr, "Processed releaseDomain, but domain %d not dead\n",
+                    domid);
+        goto out;
+    }
+
     /* if we are paused don't process anything else */
     if (xen_pause_requested)
         goto out;
